async_server.c

```diff
@@ -0,0 +1,364 @@
+      int id;
+      if (queue_pop (&cl_ctx->registry_idx, &id) != QS_SUCCESS)
+        break;
+
+      trace ("Got index from registry");
+
+      task_t *task = &cl_ctx->registry[id];
+      if (queue_pop (&mt_ctx->queue, task) != QS_SUCCESS)
+        {
+          if (queue_push (&cl_ctx->registry_idx, &id) != QS_SUCCESS)
+            error ("Could not push back id to registry indices queue");
+          break;
+        }
+
+      trace ("Got task from global queue");
+
+      cl_ctx->registry_used[id] = true;
+
+      task_t task_copy = *task;
+      task_copy.task.id = id;
+      task_copy.to = task->from;
+      task_copy.from = 0;
+
+      if (send_task (cl_ctx->socket_fd, &task_copy) == S_FAILURE)
+        {
+          error ("Could not send task to client");
+          if (queue_push_back (&mt_ctx->queue, task) != QS_SUCCESS)
+            error ("Could not push back task to global queue");
+          cl_ctx->registry_used[task->task.id] = false;
+          break;
+        }
+    }
+
+cleanup:
+  trace ("Cleaning up sender thread");
+  pthread_cleanup_pop (!0);
+
+  return (NULL);
+}
+
+static void *
+handle_clients (void *arg)
+{
+  serv_context_t *srv_ctx = *(serv_context_t **)arg;
+  mt_context_t *mt_ctx = &srv_ctx->context;
+
+  while (true)
+    {
+      acl_context_t *acl_ctx = acl_context_init (srv_ctx);
+      if (!acl_ctx)
+        break;
+
+      while (true)
+        {
+          if (srv_ctx->socket_fd < 0)
+            {
+              error ("Invalid server socket");
+              acl_context_destroy (acl_ctx);
+              return (NULL);
+            }
+
+          acl_ctx->socket_fd = accept (srv_ctx->socket_fd, NULL, NULL);
+          if (acl_ctx->socket_fd == -1)
+            {
+              error ("Could not accept new connection: %s", strerror (errno));
+              if (errno == EINVAL)
+                {
+                  acl_context_destroy (acl_ctx);
+                  return (NULL);
+                }
+              continue;
+            }
+          break;
+        }
+
+      trace ("Accepted new connection");
+
+      int option = 1;
+      setsockopt (acl_ctx->socket_fd, SOL_SOCKET, TCP_NODELAY, &option,
+                  sizeof (option));
+
+      trace ("Copied client context");
+
+      if (thread_create (&mt_ctx->thread_pool, task_sender, &acl_ctx,
+                         sizeof (acl_ctx), "async sender")
+          == S_FAILURE)
+        {
+          error ("Could not create task sender thread");
+          close_client (acl_ctx->socket_fd);
+          acl_context_destroy (acl_ctx);
+          continue;
+        }
+
+      trace ("Created a sender thread %08x",
+             mt_ctx->thread_pool.threads.prev->thread);
+
+      if (thread_create (&mt_ctx->thread_pool, result_receiver, &acl_ctx,
+                         sizeof (acl_ctx), "async receiver")
+          == S_FAILURE)
+        {
+          error ("Could not create result receiver thread");
+          close_client (acl_ctx->socket_fd);
```

Comment:
here we need to decrease ref_count and cancel "sender" thread

Reply:
Done